### ISSUE TITLE
ZFS determines an incorrect size for an ACL, resulting in a kernel panic

### ISSUE THREAD

**Issue by sadiesue43** [openzfs/zfs#7910]
### System information
Type | Version/Name
 --- | --- 
Distribution Name	| Ubuntu
Distribution Version	| 18.04
Linux Kernel	| 4.15.0-34-generic #37-Ubuntu
Architecture	| x86_64
ZFS Version	| 0.7.10-1
SPL Version	| 0.7.10-1

### Describe the problem you're observing
After upgrading an old ZPool/ZFS from Ubuntu 16.04/Linux Kernel 4.10.0, began receiving kernel panics upon access.  Through debugging, the bug has likely existed before (at least since 0.6) but the kernel wasn't specifically checking for buffer overflows beforehand.

I've been discussing my issue on zfs-discuss under the thread ['ZFS File System inaccessible after upgrade',](http://list.zfsonlinux.org/pipermail/zfs-discuss/2018-September/032181.html) but I think all relevant info is here.

### Describe how to reproduce the problem
Curate a ZFS filesystem for 10+ years from OpenSolaris through OpenIndiana, various flavors of ZFS on Linux... But seriously, I can reproduce the problem immediately on a test VM I've created from the smaller 300G filesystem; not super easy to distribute but I can definitely test any fixes necessary.

### Include any warning/errors/backtraces from the system logs
Original dmesg output:
```
[ 320.449464] detected buffer overflow in memmove
[ 320.449489] ------------[ cut here ]------------
[ 320.449492] kernel BUG at /build/linux-wuhukg/linux-4.15.0/lib/string.c:1052!
[ 320.449503] invalid opcode: 0000 [#1] SMP PTI
[ 320.449506] Modules linked in: thunderbolt nls_iso8859_1 zfs(PO) zunicode(PO) zavl(PO) icp(PO) zcommon(PO) znvpair(PO) spl(O) snd_
hda_codec_hdmi intel_rapl x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel kvm snd_hda_codec_ca0132 irqbypass crct10dif_pclmu
l crc32_pclmul ghash_clmulni_intel pcbc aesni_intel snd_hda_intel snd_hda_codec aes_x86_64 crypto_simd snd_hda_core glue_helper crypt
d intel_cstate snd_hwdep intel_rapl_perf snd_pcm intel_wmi_thunderbolt snd_seq_midi snd_seq_midi_event snd_rawmidi snd_seq snd_seq_de
vice snd_timer snd soundcore input_leds mac_hid mei_me mei shpchp acpi_pad intel_pch_thermal sch_fq_codel vhba(OE) parport_pc ppdev l
p parport ip_tables x_tables autofs4 hid_generic usbhid hid i915 mxm_wmi e1000e i2c_algo_bit ptp drm_kms_helper pps_core syscopyarea
sysfillrect
[ 320.449599] sysimgblt fb_sys_fops alx drm mdio ahci libahci wmi video
[ 320.449617] CPU: 3 PID: 3202 Comm: updatedb.mlocat Tainted: P W OE 4.15.0-32-generic #35-Ubuntu
[ 320.449621] Hardware name: Gigabyte Technology Co., Ltd. To be filled by O.E.M./Z170X-Gaming 7, BIOS F8 08/26/2016
[ 320.449632] RIP: 0010:fortify_panic+0x13/0x22
[ 320.449636] RSP: 0018:ffffa85fc5d27898 EFLAGS: 00010282
[ 320.449641] RAX: 0000000000000023 RBX: ffffa85fc5d279a8 RCX: 0000000000000000
[ 320.449645] RDX: 0000000000000000 RSI: ffff97dc51d96498 RDI: ffff97dc51d96498
[ 320.449649] RBP: ffffa85fc5d27898 R08: 0000000000000000 R09: 000000000000058c
[ 320.449652] R10: ffff97dc2916501c R11: ffffffffb895380d R12: 0000000000000001
[ 320.449656] R13: ffff97dc2a717460 R14: 0000000000000070 R15: ffff97dc21e40300
[ 320.449661] FS: 00007fb9bd9c4540(0000) GS:ffff97dc51d80000(0000) knlGS:0000000000000000
[ 320.449665] CS: 0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 320.449669] CR2: 00007fa69800e778 CR3: 0000000812c88005 CR4: 00000000003606e0
[ 320.449673] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[ 320.449677] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[ 320.449680] Call Trace:
[ 320.449779] zfs_acl_node_read.constprop.15+0x33d/0x340 [zfs]
[ 320.449867] zfs_zaccess_aces_check+0x96/0x380 [zfs]
[ 320.449874] ? mutex_lock+0x12/0x40
[ 320.449879] ? _cond_resched+0x19/0x40
[ 320.449929] ? arc_buf_access+0x14a/0x270 [zfs]
[ 320.449977] ? dbuf_rele_and_unlock+0x1a8/0x4b0 [zfs]
[ 320.450056] zfs_zaccess_common+0xda/0x1f0 [zfs]
[ 320.450062] ? _cond_resched+0x19/0x40
[ 320.450139] zfs_zaccess+0xd4/0x3d0 [zfs]
[ 320.450235] ? rrw_enter_read_impl+0xae/0x160 [zfs]
[ 320.450327] zfs_lookup+0x1c5/0x3a0 [zfs]
[ 320.450415] zpl_lookup+0xc9/0x1e0 [zfs]
[ 320.450424] lookup_slow+0xab/0x170
[ 320.450432] walk_component+0x1c3/0x470
[ 320.450438] ? _cond_resched+0x19/0x40
[ 320.450444] ? mutex_lock+0x12/0x40
[ 320.450451] path_lookupat+0x84/0x1f0
[ 320.450458] ? ___slab_alloc+0xf2/0x4b0
[ 320.450545] ? zfs_readdir+0x267/0x460 [zfs]
[ 320.450552] ? ___slab_alloc+0xf2/0x4b0
[ 320.450560] filename_lookup+0xb6/0x190
[ 320.450569] ? __check_object_size+0xaf/0x1b0
[ 320.450578] ? strncpy_from_user+0x4d/0x170
[ 320.450586] user_path_at_empty+0x36/0x40
[ 320.450592] ? user_path_at_empty+0x36/0x40
[ 320.450600] vfs_statx+0x76/0xe0
[ 320.450606] ? _cond_resched+0x19/0x40
[ 320.450611] ? dput.part.22+0x2d/0x1e0
[ 320.450619] SYSC_newlstat+0x3d/0x70
[ 320.450628] ? SyS_poll+0x9b/0x130
[ 320.450635] ? SyS_poll+0x9b/0x130
[ 320.450643] SyS_newlstat+0xe/0x10
[ 320.450651] do_syscall_64+0x73/0x130
[ 320.450659] entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[ 320.450664] RIP: 0033:0x7fb9bd4d2815
[ 320.450669] RSP: 002b:00007ffe3d613f38 EFLAGS: 00000246 ORIG_RAX: 0000000000000006
[ 320.450675] RAX: ffffffffffffffda RBX: 00005632906b5059 RCX: 00007fb9bd4d2815
[ 320.450679] RDX: 00007ffe3d613fb0 RSI: 00007ffe3d613fb0 RDI: 00005632906b5059
[ 320.450684] RBP: 00005632906ab600 R08: 0000000000000073 R09: 0000000000000073
[ 320.450688] R10: 0000000000000000 R11: 0000000000000246 R12: 00007ffe3d614190
[ 320.450692] R13: 0000000000000000 R14: 00007ffe3d614180 R15: 0000000000000000
[ 320.450697] Code: e0 4c 89 e2 e8 61 6a 00 00 42 c6 04 20 00 48 89 d8 5b 41 5c 5d c3 0f 0b 55 48 89 fe 48 c7 c7 f0 10 1a b8 48 89 e5 e8 7f cf 76 ff <0f> 0b 90 90 90 90 90 90 90 90 90 90 90 90 90 55 31 c9 48 89 fa
[ 320.450788] RIP: fortify_panic+0x13/0x22 RSP: ffffa85fc5d27898
[ 320.450794] ---[ end trace 8e2de80852d9eb6a ]---
```

Patched zfs_acl.c with this code to find more information:
```
diff --git a/module/zfs/zfs_acl.c b/module/zfs/zfs_acl.c
index 1fcfca0c7..2bc614293 100644
--- a/module/zfs/zfs_acl.c
+++ b/module/zfs/zfs_acl.c
@@ -1107,8 +1107,13 @@ zfs_acl_node_read(znode_t *zp, boolean_t have_lock, zfs_acl_t **aclpp,
                            znode_acl.z_acl_extern_obj, 0, aclnode->z_size,
                            aclnode->z_acldata, DMU_READ_PREFETCH);
                } else {
+                       int copysize = aclnode->z_size;
+                       if (copysize > sizeof(znode_acl.z_ace_data)) {
+                               printk(KERN_NOTICE "ZFS: PANIC: zfs_acl_node_read acl size too large for objid %llu: trying to copy %d bytes to %ld sized z_ace_data. Limiting data copied.\n", zp->z_id, copysize, sizeof(znode_acl.z_ace_data));
+                               copysize = sizeof(znode_acl.z_ace_data);
+                       }
                        bcopy(znode_acl.z_ace_data, aclnode->z_acldata,
-                           aclnode->z_size);
+                           copysize);
                }
        } else {
                error = sa_lookup(zp->z_sa_hdl, SA_ZPL_DACL_ACES(ZTOZSB(zp)),
```

Additional dmesg output with debug code in place:
```
[   68.993450] ZFS: PANIC: zfs_acl_node_read acl size too large for objid 6: trying to copy 112 bytes to 72 sized z_ace_data. Limiting data copied.
[   68.993607] ZFS: PANIC: zfs_acl_node_read acl size too large for objid 15493: trying to copy 112 bytes to 72 sized z_ace_data. Limiting data copied.
```

**Comment by loli10K** [comment]
Duplicate of #7581

**Comment by sadiesue43** [comment]
Hm, a 3.5 month old open bug with no comments doesn't bode well for me, does it?  :)

I'll see what I can do about reproducing @enzinol bug and try to figure out where the oversized ACLs are being created.

**Comment by enzinol** [comment]
I am open to workarounds as well... 👍

**Comment by siebenmann** [comment]
@enzinol: If you patch the 0.7.x code with the patch here, you should get kernel warning messages but no panics. The files and objects should be accessible, at least enough to read them out and back them up, although since we are now only copying some of the ACL data there may be other issues.

So that I don't forget: the message in the patch here is actually incorrect, which is my fault; I misread the order of arguments in `bcopy()`. We are trying to copy 112 bytes from a 72-byte `z_ace_data` to a 112-byte ACL buffer; if the kernel let us do this, we would be copying an extra 40 bytes of random garbage from beyond the `z_ace_data` (which is at the end of its structure).

@sadiesue43: Given this, it would be interesting to augment the message to also log the `acl_count` variable, since I think that is theoretically how many ACL entries there are. It would also be interesting to 
log `znode_acl.z_acl_version`, because I suddenly have a suspicion here, which I might as well write down in this comment.

The ACL size here is determined by `zfs_acl_znode_info()`. This is not a `z_is_sa` ACL, so there are two cases; old ACLs (`ZFS_ACL_VERSION_INITIAL`) and new ones. New ACLs have their size and count recorded directly; old ACLs have their count in what is now the size field, and their size is determined through `ZFS_ACL_SIZE()`, which is a macro that multiplies the count by what is supposed to be the size of a single ACL structure. Based on types in include/sys/zfs_acl.h, it appears that the physical old ACEs are `zfs_oldace_t` (and this is the definition used to determine the size of `z_ace_data`, see `ZFS_ACE_SPACE()`). However, `ZFS_ACL_SIZE()` uses a completely different type to determine the size of data to copy, `ace_t`, which comes from include/spl/sys/acl.h. I cannot tell if `ace_t` is the same size as a `zfs_oldace_t`, and in any case it seems like this is at least questionable.

This type confusion is in the upstream ZFS source code, but an `ace_t` likely means something different in Illumos. In ZFS on Linux, `ZFS_ACL_SIZE()` is only used in this one spot.

(However, if `z_acl_version` is not '0' then something else is going on.)

**Comment by sadiesue43** [comment]
New patched line:
```
printk(KERN_ERR "ZFS: PANIC: zfs_acl_node_read acl size too large for objid %llu: trying to copy %d bytes to %ld sized z_ace_data. Limiting data copied. ACL_count %d, ACL_VERSION %d\n", zp->z_id, copysize, sizeof(znode_acl.z_ace_data), acl_count, znode_acl.z_acl_version);
```

New dmesg output:
```
[  119.493144] ZFS: PANIC: zfs_acl_node_read acl size too large for objid 6: trying to copy 112 bytes to 72 sized z_ace_data. Limiting data copied. ACL_count 10, ACL_VERSION 1
[  119.493454] ZFS: PANIC: zfs_acl_node_read acl size too large for objid 15493: trying to copy 112 bytes to 72 sized z_ace_data. Limiting data copied. ACL_count 10, ACL_VERSION 1
```

**Comment by enzinol** [comment]
I did look for ACLs on files and directories (below directory only), but all my files and directories have only 3 ACLs.
`find . -type d | while read filename; do count=$(getfacl "$filename" | grep "::" -c); echo $filename $count; done | grep -v " 3$"`

I will try the patch above at one point.

**Comment by siebenmann** [comment]
If I'm correct in my understanding of things, filesystems with this problem have 'xattr=sa' and 'acltype=posixacl', but it would be useful to verify that just to be sure. Also, is the `dnodesize` property set to 'legacy' or something else?

I believe that you can dump the raw SA xattrs (with some size information about them) with zdb. For me, with a specific object ID, this looks like:

````
# zdb -vvv -ddddd tank/acltest 129
[...]
        SA xattrs: 108 bytes, 1 entries

                system.posix_acl_access = \002\000\000\000\001\000\006\000\377\377\377\377\002\000\004\000s\010\000\000\004\000\000\000\377\377\377\377\020\000\004\000\377\377\377\377 \000\000\000\377\377\377\377
[...]
````

I don't know if this information can be decoded to anything useful, but maybe it can be.

(The 'SA xattrs' count isn't how many ACL entries there are, it's how many extended system attributes there are; here there's one, for the ACL. I think we would have to decode the raw bytes to a structure in order to know how many ACLs there were. But I would really start having to read the code here.)

**Comment by enzinol** [comment]
Not in my case:
`NAME         PROPERTY  VALUE     SOURCE
Pool/Photos  acltype   off       default`

**Comment by siebenmann** [comment]
@enzinol That's what they call very interesting, because this crash is definitely trying to read an ACL and it thinks it has one. If you could find a specific file that causes the problem, a zdb dump of it would be pretty interesting. You can tell zdb to dump a specific path (instead of an object ID) with the `-O` option, eg `zdb -vvv -ddddd -O Pool/Photos some/file`.

(Is it possible that your filesystem had ACLs on at one point but had them turned off since then?)

**Comment by enzinol** [comment]
At this point patching in order to get which object/file causes the issue (I am running 16.04 because I cannot upgrade due to this issue) is required. I will attempt to build the latest ZOL from sources and report back. Hopefully @sadiesue43 can help shed a light.

**Comment by siebenmann** [comment]
This ACL code hasn't changed in quite a while, so it's quite possible that the patch will apply cleanly to your older 0.6.x code as well. If you test on 16.04 with patched code, you should get the same log messages (and the amount of ACL data copied will be limited, so you might want to mount things read-only while you're testing).

**Comment by sadiesue43** [comment]
Sorry for the delayed response.

It seems my data is defaulted as well for all ZFS:
PROPERTY              VALUE                  SOURCE
xattr                 on                     default
acltype               off                    default
dnodesize             legacy                 default

However, those values may have changed in the ~10 years I've had this pool (not likely by me, but possibly if a default has changed since the OpenSolaris days).

I've attached my [ZDB](https://github.com/zfsonlinux/zfs/files/2392814/zdb.txt) output.

I might be able to create a ~100MiB sample snapshot send, and will get that uploaded if it is successful.

**Comment by sadiesue43** [comment]
Here is a [sample](https://github.com/zfsonlinux/zfs/files/2393012/steve.zip) zfs snapshot send with this issue.  I took my working test dataset (~300G), deleted everything but object 15493 (from dmesg failures), then truncated that .exe file to 5M (so I could get under GitHub's upload limits).  I'm able to zfs recv this snapshot and cause the bug to manifest; hopefully it will be of use to you @siebenmann !

**Comment by siebenmann** [comment]
I'm going to take a look at this when I have time, but right away things are different from what I expected and what the code suggests. There are two ways for ACLs to be stored in ZFS, as system attributes of dnodes and as separate objects (this is what the `xattr` property controls). From the code I was reading, it appeared that you had SA ACLs, which would have produced `zdb` output similar to what I showed in my example. Instead you have external object ACLs, in the `xattr` object IDs 15494 and 15495, which it might be worthwhile to dump with `zdb` to see if it reports anything on them.

**Comment by sadiesue43** [comment]
[15494](https://github.com/zfsonlinux/zfs/files/2393677/15494.txt) does not exist in the subset snapshot - it was for another top level directory. 
[15495](https://github.com/zfsonlinux/zfs/files/2393667/15495.txt) is the same between the original dataset and the sub-dataset.

**Comment by siebenmann** [comment]
Both of these are xattr directories, as expected. 15494 is empty, which is interesting; 15495 contains what is probably something to do with Solaris's SMB server, given that 'SUMWsmb:...' identifier. Dumping object 15496 will probably be the end of the chain. But if the subset snapshot reproduces the issue on both things, and 15494 isn't included in the subset snapshot, then that's interesting all by itself.

(I'm not sure why it would be excluded if the '/WinXP/' directory was included, but the ways of ZFS are mysterious. Maybe it noticed that an empty xattr directory was surplus when generating the snapshot.)

**Comment by enzinol** [comment]
Interesting. I don't recall if mine has roots in OpenIndiana, it's very likely as I don't remember creating the pool on Linux directly.
I assume that I can get away by moving the content to a different dataset, right?

**Comment by sadiesue43** [comment]
[15496.txt](https://github.com/zfsonlinux/zfs/files/2393797/15496.txt)

I may not have been clear on the subset snapshot.  I deleted everything on the dataset except a single .exe file at the top level, and then had to truncate that .exe as well.  ZFS Recv that subset and I can hit the bug with object ID 15493 that I also hit with the full dataset (though I suspect ObjID 6 was overwritten through the deletes/truncation).  WinXP is definitely one of the directories I deleted.

**Comment by enzinol** [comment]
My quick attempt to build from sources and install on top of my 16.04 installation didn't work (I didn't see the log message while importing the pool). My suspicion is that I messed up with some dependencies. Next attempt: I will try to move the files to new datasets and try a to import the pool from a live USB install...

**Comment by siebenmann** [comment]
You can check what ZFS module version and so on you're getting by looking at `dmesg` output or eg `modinfo zfs`, although that only helps if your from-source build has a different version than the regular one. Also, I would expect this panic/log message to trigger only when ZFS looks at an affected file or directory, although that might happen during pool import depending on what's going on. If you want to see for sure, import the pool, mount all the filesystems, and then do something that stat's all of the files (the brute force way is probably `ls -lR .... >/dev/null`).

**Comment by enzinol** [comment]
> You can check what ZFS module version and so on you're getting by looking at dmesg output or eg modinfo zfs, although that only helps if your from-source build has a different version than the regular one.

I will add a custom version string to verify that it is indeed the correct binary.

> Also, I would expect this panic/log message to trigger only when ZFS looks at an affected file or directory, although that might happen during pool import depending on what's going on. 

My original bug linked was triggered during import (or mounting?) so I expect the deadly ACL to be attached to a very high level object (maybe one of the file systems themselves?). I will keep you posted.

**Comment by siebenmann** [comment]
Equipped with @sadiesue43 's snapshot, I was able to reproduce this in 0.7.11 (on top of Fedora 28) and then poke through the ZFS filesystem directly. I ended up hacking together a version of zdb that directly dumped the on-disk ACL data, which confirms that it's incorrect on disk. For a good ACL, such as the ACL on / (the root directory of the filesystem, objid 3) it reports:

````
        acl     objid 0 size 48 version 1 count 6
        ace_data = 00000010bf011e0000 004020af00120000 000040af00120000 0040202700000001 00004017010c0000 000040a8001200
````

For the bad file (objid 15493) this reports:
````
        acl     objid 0 size 112 version 1 count 10
        ace_data = TRUNCATED TO 72 BYTES: 00000010bf011e0000 004020a900120000 000040a900120012 0000000100000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 00000000000000
````

An objid of 0 here (representing the field `z_acl_extern_obj`) means that the ACL data is supposed to be embedded in the `z_ace_data` portion of the ACL structure. But that portion is only 72 bytes long, while the size of the ACL is claimed to be 112 bytes. In addition, many of the trailing bytes (and thus the trailing ACE structures) look like they're just zeroed and thus are effectively bogus (I think technically they can be interpreted as a type of ACE, but).

So it looks like there are two problems here. First, you really do have bad data on disk; as far as I can tell, it's not that the ZFS code is mis-interpreting some acceptable data. This bad data has probably been there for a very long time and has been mostly harmless until now. Second, the Linux kernel is now detecting that ZFS on Linux is doing a bad memory copy operation as a result of this bad (on-disk) data, and when it detects this you get the panic. The good news is that the memory copy is not overwriting data it shouldn't be, it's simply reading (and copying) bad data.

Since the bad data exists on disk, it's my view that ZFS shouldn't panic when it runs into it. Since we have at least two people with this problem, it's probably worth doing something to prevent the panic, although I'm not sure what should be done here. Likely the right answer is to bail out of `zfs_acl_node_read()` with some form of error, instead of trying to continue to interpret a corrupted ACL.

For interested parties, you can find my modified 0.7.11 zdb.c [here](https://www.cs.toronto.edu/~cks/vendors/zfs-on-linux/issue-7910), both in full form and as a patch. The patch applies to the 18.04 zfsutils-linux source package, and you can compile that by hand to get a new zdb binary. The patch appears to apply cleanly to the 0.6.x version that's in Ubuntu 16.04, so I think that you could rebuild a modified zdb even there and use it to check suspect files or even entire filesystems. To check an entire filesystem, use `zdb -vvv -dddd pool/filesystem >/tmp/output` (this will take a while) and then look through the output for `TRUNCATED TO` warnings.

**Comment by enzinol** [comment]
Thanks @siebenmann, I will build zdb later and try with the remaining file systems.
In the meanwhile I am in a zfs rename, zfs create, mv -v loop
(unfortunately I discovered that mv doesn't delete the files on the go which is terrible as I have a few filesystems that are larger than the free space in the Pool).
I agree on the part that the file system should be resilient to this and avoid panicking. 
One question I have: is z_ace_data size computed or fixed? If it's fixed, I am guessing truncation would be probably fine as clearly the ACL on disk size is wrong?
If it's computed, what is the root of the discrepancy?

**Comment by sadiesue43** [comment]
Thanks again @siebenmann , I think I'll also begin the "copy/paste for healing" loop @enzinol has begun.  Probably for the best with how old this pool is anyway.  :)  And likely no data loss is definitely a huge plus!

**Comment by siebenmann** [comment]
`z_ace_data` is a fixed size thing; it's always 72 bytes, apparently intended and designed to hold up to 6 standard-sized 'ACE' entries embedded in the overall on-disk ACL description structure (I suspect that ACE stands for Access Control Entry). The curious can see the definitions for it in [zfs_acl.h](https://github.com/zfsonlinux/zfs/blob/zfs-0.7-release/include/sys/zfs_acl.h); you want to look at `ZFS_ACE_SPACE` and `zfs_acl_phys_t`. The original ACEs were apparently always 12 bytes long (in two 32-bit ints and two 16-bit ints), so 6 of them come out to 72 bytes; later ACEs are apparently variable-sized but always at least 8 bytes.

I suspect that truncation is fine, but returning an error is probably more correct (and thus perhaps more likely to be accepted as a change). Returning an error makes sure that nothing can misinterpret the truncated results.

**Comment by enzinol** [comment]
Thank @siebenmann  for your analysis!
@sadiesue43 thank you for opening this bug and starting to dig. Hopefully the workaround works and this weekend I can attempt to upgrade to 18.04.1

**Comment by enzinol** [comment]
Another idea: it would be nice if this issue would popup as a warning during a scrub.

**Comment by loli10K** [comment]
> (I'm not sure that scrubs even parse ZFS filesystem directories at the moment, or if they simply read all allocated ZFS inodes without caring what they are and how they're interconnected.)

AFAIK scrubs do not (yet) detect this kind of issues, see https://github.com/zfsonlinux/zfs/issues/4083.

**Comment by enzinol** [comment]
Mmm. I did "cleanse" all the files by moving them to a different dataset. Still segfault when I import the pool with the new kernel. 
Is there any way to recursively whack all the ACLs? I am starting to panic...

**Comment by enzinol** [comment]
Indeed I did a user level copy using `mv`. The only thing left is a pool root snapshot which I am afraid to destroy. Next attempt build and patch in order to id the object.

**Comment by enzinol** [comment]
What are the best instructions to build the ZFS debs for Ubuntu?
When I follow these [here](https://www.cmiss.org/cmgui/wiki/BuildingUbuntuPackagesFromSource) the compilation succeeds even if I remove the zfs_acl.c file (or I put garbage in it).

The ones [here](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS) don't work on the deb part (complains about rpm-common being empty). If I do `make install` the binaries don't work as they are missing some relevant dependencies.

Alternatively if you somebody could drop the patched debs somewhere it would be great

**Comment by siebenmann** [comment]
Ubuntu makes this unusually difficult, because the ZFS module source code is actually part of their kernel package (!); the ZFS packages they ship separately are just the user tools. The wiki page I think you want is [Custom Packages](https://github.com/zfsonlinux/zfs/wiki/Custom-Packages), but it is not 100% correct at the moment. When I built packages for Ubuntu, I used the official source and built DKMS based packages which works but isn't documented. There are two magic things you need to do differently than the instructions to get this to work. First, when you do the `make` for both SPL and ZFS, use `make; make pkg-utils deb-dkms` (the deb-dkms bit is the magic to ask for DKMS based packages). Second, when you are configuring ZFS, do `./configure --with-config=user` instead of the `./configure` the instructions tell you to run.

To install these DKMS-based packages, you'll need to install dkms itself with `apt-get install dkms`. Then you can install the .debs that the instructions tell you to. You don't necessarily want to install all of them, but I have lost track of which ones are necessary and which ones aren't. Installing the packages built by hand this way will cause your system to diverge from Ubuntu packaging (even though you're installing through .deb packages); getting back to a stock setup is something I've never had to figure out. You may also have to remove some of the stock Ubuntu ZFS packages first due to file conflicts, although I'm not certain; I did my work on a system without the Ubuntu ZFS packages installed.

(I think you want both spl*_amd64.deb debs and then libnvpair1, libuutil1, libzfs2, libzpool2, zfs, and zfs-dkms; I would install them in more or less that order.)

You will want to make sure that dkms installed both SPL and ZFS modules with `dkms status`, and then reboot to activate the right kernel modules. Once you've installed the DKMS packages, you can directly modify the source of the ZFS kernel modules in /usr/src/zfs-0.7.11/module/zfs (or whatever version) and then force dkms to rebuild and reinstall them without going all the way back to rebuilding and reinstalling the .debs. This is not clean, but it is convenient for hacking and testing (ie, it's a fair bit faster).

**Comment by enzinol** [comment]
Thanks. I don't care much about diverging, my aim is to figure out the offending object and go from there. I am using a live distro for now, considering switching to a more hacking and testing friendly one as an alternative.

**Comment by siebenmann** [comment]
If you can use any distro, I would seriously consider using a distribution that uses (or supports) DKMS based ZFS packages, either as part of the distro or by ZFS on Linux. Based on looking at the ZoL site and its links for distros, the ones with DKMS are Debian, RHEL/CentOS, and Fedora (which I have the most experience with). The advantage of using one of these with DKMS is that you don't have to rebuild the packages themselves; you just install the official packages, edit /usr/src/zfs-0.7.x/module/zfs/zfs_acl.c, and force DKMS to rebuild and reinstall the kernel modules. I know this definitely works for Fedora (I did testing this way), and I assume it works just the same for the other two.

(The one trick with doing things this way is always look in /var/lib/dkms/zfs/0.7.x/$(uname -r)/x86_64/log/make.log for warning messages from the compiler or outright reports of syntax errors. I almost blew my foot off several times this way, for example by using the wrong printf() format specifiers.)

**Comment by enzinol** [comment]
Debian it is! I will keep you posted.

**Comment by enzinol** [comment]
Thanks @gdevenyi 
Do you have pointers on how to replace standard kernel modules with DKMS? That seems the quickest way as I already have a working USB live device.

**Comment by gdevenyi** [comment]
@siebenmann above notes the dkms of Ubuntu is weird and hard to use, but he provides some instructions

**Comment by gdevenyi** [comment]
Oh my bad, just install the dkms packages and then edit according to previous instructions

**Comment by gdevenyi** [comment]
https://packages.ubuntu.com/bionic/zfs-dkms

**Comment by kpande** [comment]
please use the 'edit' button instead of adding so many comments.

**Comment by enzinol** [comment]
Bingo.
`[    9.470018] ZFS: PANIC: zfs_acl_node_read acl size too large for objid 3: trying to copy 304 bytes to 72 sized z_ace_data. Limiting data copied. ACL_count 23, ACL_VERSION 1`

How do I know to which dataset it belongs?
The output of `zdb -vvv -ddddd Pool 3` is
`Dataset mos [META], ID 0, cr_txg 4, 41.1M, 306 objects, rootbp DVA[0]=<0:640001cc000:2000> DVA[1]=<0:8623670a000:2000> DVA[2]=<0:2a000fac000:2000> [L0 DMU objset] fletcher4 uncompressed LE contiguous unique triple size=800L/800P birth=12284894L/12284894P fill=306 cksum=40b10875d:77cc73aaf05:6f31d0fec1584:44f92e589068791

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         3    1    16K    16K  35.0K     512    32K  100.00  DSL props (K=inherit) (Z=inherit)
	dnode flags: USED_BYTES 
	dnode maxblkid: 1
	Fat ZAP stats:
		Pointer table:
			1024 elements
			zt_blk: 0
			zt_numblks: 0
			zt_shift: 10
			zt_blks_copied: 0
			zt_nextblk: 0
		ZAP entries: 3
		Leaf blocks: 1
		Total blocks: 2
		zap_block_type: 0x8000000000000001
		zap_magic: 0x2f52ab2ab
		zap_salt: 0x76b1dfe27
		Leafs with 2^n pointers:
			  9:      1 *
		Blocks with n*5 entries:
			  0:      1 *
		Blocks n/10 full:
			  1:      1 *
		Entries with n chunks:
			  3:      2 **
			  4:      1 *
		Buckets with n entries:
			  0:    509 ****************************************
			  1:      3 *

		$hasrecvd = 0 
		sharesmb$recvd = off
		nms:dedup-dirty$recvd = off Indirect blocks:
               0 L0 0:c4000:2000 0:2c0000c4000:2000 0:5800006c000:2000 4000L/1000P F=1 B=77/77
            4000 L0 0:c6000:2000 0:2c0000c6000:2000 0:5800006e000:2000 4000L/1000P F=1 B=77/77

		segment [0000000000000000, 0000000000008000) size   32K

**Comment by enzinol** [comment]
@siebenmann, @loli10K  if I am reading this correctly, the objectId in question is a pool property node. In this case

1. why would it have an ACL?
2. how do I clear it or fix it?

**Comment by siebenmann** [comment]
I think that the only things that have ACLs are things in ZFS filesystems, so it will be object id 3 in some filesystem (the message doesn't print any sort of filesystem ID for various reasons, including that I didn't see an obvious way to easily get one). Since you experience a panic on pool import and this is such a low object ID, my suspicion is that this is the root of some filesystem; that would explain why merely importing the pool (which mounts all of the filesystems) would provoke the panic without you specifically accessing anything.

To find the filesystem, I can think of a number of options. First, you can try to do `ls -ld /pool/fs/.` for every filesystem, one by one, while watching the kernel messages. If you're lucky, accessing one of them will trigger this message and then you have your victim. Failing that, try `zfs unmount -a` and then `zfs mount` each filesystem one by one, again while watching the kernel logs to see when the message shows up. And finally you can build my patched `zdb`, then use it on object ID 3 in each filesystem. The one with the problem will print out that 'TRUNCATED TO' bit about its ACL.

As for how you clear the ACL once you've found the filesystem ... that is an excellent question that I don't know the answer to. Sorry.

**Comment by enzinol** [comment]
Ok. I did try to dump the objectId 3 for all the datasets but they were all empty.
**Good news**: playing with attributes, I guess setting `xattr=sa`, **AND** rebooting seems to have fixed the issue. I accidentally discovered this while backing up the entire pool to a new set of HDs, which was something that was overdue anyway.
Once the backup is done, I will check with the standard kernel modules to see if I am hallucinating or not. If not, I think I am done and finally upgrade to the latest and greatest. Otherwise I will try to mount each file-system separately to find the culprit.
 Thanks!!!

**Comment by enzinol** [comment]
Thanks, I can confirm that my pool now is working.
Let's discuss now what is the most appropriate way to fix this. :+1:
